server/topic.go

```diff
@@ -3962,3 +3975,35 @@ func topicNameForUser(name string, uid types.Uid, isChan bool) string {
 	}
 	return name
 }
+
+// calculateUnreadInRanges calculates how many unread messages are within the given ranges.
+// unreadStart is the first unread message SeqId (readID + 1), unreadEnd is the last possible message SeqId.
+func calculateUnreadInRanges(readID, lastID int, ranges []types.Range) int {
+	if readID >= lastID {
+		// No unread messages
+		return 0
+	}
+	
+	unreadStart := readID + 1
+	unreadEnd := lastID
+	
+	count := 0
+	for _, r := range ranges {
```

Comment:
This code seems to be very inefficient. First, the ranges are sorted, second, you need to find the ranges where rangeEnd > readID.